tc/core/compiler-inl.h

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "tc/core/compiler.h"
+
+#include <sstream>
+#include <string>
+
+#include "tc/core/flags.h"
+#include "tc/core/halide_utils.h"
+#include "tc/core/tensor.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace {
+void checkInputsCompliant(
+    const tc2halide::HalideComponents& halideComponents,
+    const std::vector<const DLConstTensor*>& inputsInfo) {
+  if (inputsInfo.size() != halideComponents.inputs.size()) {
+    throw lang::ErrorReport(halideComponents.getDef())
+        << "expected " << halideComponents.inputs.size() << " inputs but found "
+        << inputsInfo.size();
+  }
+  for (size_t i = 0; i < inputsInfo.size(); ++i) {
+    auto dltype = inputsInfo[i]->dtype;
+    auto htype = halideComponents.inputs[i].type();
+    // we have three type representations here: (1) halide Type (2) DLTensor
+    // type, and (3) the token representing the type in the frontend (e.g.
+    // TK_FLOAT) we need to translate to (3) to report user facing errors
+    auto dltype =
+        lang::TypeInfo(lang::TypeInfo::Code(dltype.code), dltype.bits)
+            .toScalarToken();
+    auto htype =
+        lang::TypeInfo(lang::TypeInfo::Code(htype.code()), htype.bits())
+            .toScalarToken();
+    if (dltype != htype) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected type " << lang::kindToString(htype) << " but found "
+          << lang::kindToString(dltype);
+    }
+    int hdim = halideComponents.inputs[i].dimensions();
+    int dldim = inputsInfo[i]->ndim;
+    if (dldim != hdim) {
+      throw lang::ErrorReport(halideComponents.getDef().params()[i])
+          << "expected a tensor with " << hdim << " dimensions but found "
+          << dldim << " dimensions.";
+    }
+  }
+}
+} // namespace
+template <typename Backend>
+std::unique_ptr<typename Backend::ExecutorType> compile(
+    const std::string& tc,
+    const std::string& entryPoint,
+    const std::vector<const DLConstTensor*>& inputs,
+    /* TODO: in the future also pass outputs for stride and alignment info */
+    const typename Backend::MappingOptionsType& options) {
+  auto parsedTcs = detail::parse(tc);
+  CHECK_EQ(1, parsedTcs.count(entryPoint))
```

Comment:
Isn't this generating a warning for you?
I remember having to change all these constants to be unsigned.
Also, I find it much easier to read if you put the target of the check
before what you are checking it against.

Same for the other `CHECK_EQ` introduced in this commit.

Reply:
Old, habit, it used to be (expected, actual).

Reg warnings .. no it's explicitly disabled on my machine due to [this](https://github.com/facebookresearch/TensorComprehensions/blob/master/CMakeLists.txt#L23)
```
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-variable -Werror -Wno-sign-compare")
```

I disabled locally, I checked I indeed compile with -Wall -Werror but I still do not see anything.
I'll still make it 1u